pageserver/src/virtual_file.rs

```diff
@@ -0,0 +1,221 @@
+//!
+//! VirtualFile is like a normal File, but it's not bound directly to
+//! a file descriptor. Instead, the file is opened when it's read from,
+//! and if too many files are open globally in the system, least-recently
+//! used ones are closed.
+//!
+//! To track which files have been recently used, we the clock algorithm
+//! and a 'recently_used' flag on each slot.
+//!
+//! This is similar to PostgreSQL's vfd.c facility.
+//!
+use lazy_static::lazy_static;
+use std::fs::File;
+use std::io::{Error, ErrorKind, Read, Seek, SeekFrom};
+use std::os::unix::prelude::FileExt;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::RwLock;
+
+// TODO: probe the actual system limit, or at least make this configurable.
+const MAX_OPEN_FILES: usize = 100;
+
+pub struct VirtualFile {
+    /// Lazy handle to the global file descriptor cache. The slot that this points to
+    /// might contain our File, or it may be empty, or it may contain a File that
+    /// belongs to a different VirtualFile.
+    handle: RwLock<SlotHandle>,
```

Comment:
Could you please clarify why do we need `RwLock` here? In my understanding different threads can concurrently access one `Slot`, not `SlotHandle`. If we want to have multiple VirtualFile's pointing to the same SlotHandle it will require Rc/Arc. Please correct me if I'm wrong.

Reply:
No, multiple VirtualFiles cannot have the same SlotHandle. But multiple threads can access the underlying file through a shared `&VirtualFile` reference, just like with `&File`. We need the RwLock so that if two threads try to read (or write) the file at the same time, using the same VirtualFile, only one thread will try to open the file and the other one will wait.

Reply:
> just like with &File

I see `impl Write for File` which contains `fn write(&mut self, buf: &[u8]) -> Result<usize>`. Note that it requires `&mut self`. So I do not see how it is possible to write to several &File concurrently from different threads directly. (Excluding into_raw_fd/from_raw_fd unsafe trickery and opening the same file from different threads).

I suggest to cover shared VirtualFiles with Arc/Mutex. So VirtualFile owns its SlotHandle, and if someone owns VIrtualFile there is no need to use locks.

Reply:
See `FileExt::write_at`

Reply:
I wonder why do we need to implement non0reentrant Write trait for VirtualVile at all?  We can always use read_at/write_at...

Reply:
True. Some code uses the Read/Write traits, also in external crates (e.g. serde), but we could provide an adapter struct that implements Read/Write on top of VirtualFile.

That said, it's also nice that VirtualFile is as close as possible to std::fs::File, so that it can be used as a drop-in replacement.

Reply:
> See FileExt::write_at

Oh, didn't see that. Good to know, thanks! 
Could you please add an example/test with concurrent access? So one can clearly see how it's possible with VirtualFile. Also do we have that case in our codebase (or where is it supposed to be used)? As far as I can see now we use only owned VirtualFile. So it looks like we are paying for flexibility we do not currently use. Please correct me if I misunderstood something.

Reply:
> 
> Hmm, you mean with actually accessing a file from two threads? That seems a bit complicated. I'm inclined to just trust the rust borrow checker here.

As far as FileExt methods are read-only, Rust doesn't prevent to call them from multiple threads.
I attach my test of multithreaded IO in Rust. May be you will want ti create test for VirtualFile from it.
[rust_io.zip](https://github.com/zenithdb/zenith/files/7513885/rust_io.zip)

Reply:
Sorry, I'm a little bit late, but nevertheless let it be here :)

> Hmm, you mean with actually accessing a file from two threads?

I think about it rather as a documentation. And as a smoke test too. Borrow checker cannot reason about rwlocks with different thread interleavings. Something like [loom](https://github.com/tokio-rs/loom) might reason about this better, but using it here looks like an overkill to me.

> BookFile uses the FileExt routines, with its read_chapter function

Yeah, but everywhere except this case with `inner` protected by a mutex we have an owned book nearby. That answers this question :) So we indeed need it for ~current code.